File: 2024/day7/day7.py
import time
def prod(list_of_nums):
    if list_of_nums == []:
        return 1
    return list_of_nums[0]*prod(list_of_nums[1:])

def concat(num1, num2):
    return int(str(num1)+str(num2))

def deconcat(num1, num2):
    str1 = str(num1)
    str2 = str(num2)
    l1 = len(str1)
    l2 = len(str2)
    if l1<l2:
        return '-1'
    elif str1[l1-l2:] == str2:
        return str1[:l1-l2]
    else:
        return '-1'

def computes(num, list_of_nums):
    if num != int(num) or num<0:
        return False
    elif len(list_of_nums) == 1:
        return (num == list_of_nums[0])
    else:
        last = list_of_nums[-1]
        removed = deconcat(int(num), last)
        if removed == '':
            return True
        else:
            removed = int(removed)
        return (computes(num/last, list_of_nums[:-1])) or (computes(num-last, list_of_nums[:-1])) or (computes(removed, list_of_nums[:-1]))

def all_possible(list_of_nums):
    if len(list_of_nums) == 1:
        return list_of_nums
    else:
        last = list_of_nums[-1]
        return [x+last for x in all_possible(list_of_nums[:-1])]+[x*last for x in all_possible(list_of_nums[:-1])]+[concat(x, last) for x in all_possible(list_of_nums[:-1])]

def pruned_possible(target, nums):
    if nums == []:
        return False
    else:
        possibilities = [nums[0]]
        for i in range(1, len(nums)):
            smaller_sums = [x+nums[i] for x in possibilities if x+nums[i]<= target]
            smaller_prods = [x*nums[i] for x in possibilities if x*nums[i]<= target]
            smaller_concats = [concat(x, nums[i]) for x in possibilities if concat(x, nums[i])<=target]
            possibilities = smaller_sums+smaller_prods+smaller_concats
        if target in possibilities:
            return True
        else:
            return False

with open("day7.txt", "r") as f:
    lines = f.read().split("\n")


to_test = [[int(x.split(":")[0]), [int(y) for y in (x.split(":")[1]).split(" ")[1:]]] for x in lines]

start1 = time.time()
does_compute1 = [x[0] for x in to_test if computes(x[0], x[1])]
print("first", time.time()-start1)

start2 = time.time()
does_compute2 = [x[0] for x in to_test if x[0] in all_possible(x[1])]
print("second", time.time()-start2)

start3 = time.time()
does_compute3 = [x[0] for x in to_test if pruned_possible(x[0], x[1])]
print("does_compute3", time.time()-start3)

print(sum(does_compute1))
print(sum(does_compute2))
print(sum(does_compute3))

File: 2024/day7/day7.hs
import System.IO

parse :: String->(Int, [Int])
parse string = (head aux, tail aux)
    where
        aux = map (read::String->Int) ([takeWhile (\x-> (x/= ':')) string] ++ tail (words (dropWhile (\x->(x/=':')) string)))

deconcat :: Int->Int->String
deconcat num1 num2 = 
    let
        str1 = show num1
        str2 = show num2
        l1 = length str1
        l2 = length str2
    in
        if l1<l2 then "-1"
        else
            if (drop (l1-l2) str1)==str2 then take (l1-l2) str1
            else "-1"

computes :: (Int, [Int]) -> Bool
computes (num, list_of_nums) = 
    if (num<0) then False
    else if ((length list_of_nums) == 1) then (num == (list_of_nums!!0))
    else
        let
            l = length list_of_nums
            last = list_of_nums !! (l-1)
            all_but_last = take (l-1) list_of_nums
            removed = deconcat num last
            quotient = if (num`mod`last == 0) then (num`div`last) else -1
        in
            if removed == "" then True
            else
                let new_num = (read::String->Int) removed
                in
                    (computes (quotient, all_but_last))||(computes ((num-last), all_but_last))||(computes (new_num, all_but_last))

main = do
    handle <- openFile "day7.txt" ReadMode
    contents <- hGetContents handle
    let
        all_lines = sum [fst (parse x)|x<-(lines contents), computes (parse x)]
    print all_lines
